pvlib/test/test_numerical_precision.py

```diff
@@ -0,0 +1,108 @@
+    I_L_ref=SPR_E20_327.I_L_ref, I_o_ref=SPR_E20_327.I_o_ref,
+    R_sh_ref=SPR_E20_327.R_sh_ref, R_s=SPR_E20_327.R_s,
+    EgRef=1.121, dEgdT=-0.0002677
+)
+IL, I0, RS, RSH, NNSVTH = ARGS
+IVCURVE_NPTS = 100
+
+try:
+    from sympy import symbols, exp as sy_exp
+except ImportError as exc:
+    LOGGER.exception(exc)
+    symbols = NotImplemented
+    sy_exp = NotImplemented
+
+
+def generate_numerical_precision():
+    """
+    Generate expected data with infinite numerical precision using SymPy.
+    :return: dataframe of expected values
+    """
+    if symbols is NotImplemented:
+        LOGGER.critical("SymPy is required to generate expected data.")
+        raise ImportError("could not import sympy")
+    il, io, rs, rsh, nnsvt, vd = symbols('il, io, rs, rsh, nnsvt, vd')
+    a = sy_exp(vd / nnsvt)
+    b = 1.0 / rsh
+    i = il - io * (a - 1.0) - vd * b
+    v = vd - i * rs
+    c = io * a / nnsvt
+    grad_i = - c - b  # di/dvd
+    grad_v = 1.0 - grad_i * rs  # dv/dvd
+    # dp/dv = d(iv)/dv = v * di/dv + i
+    grad = grad_i / grad_v  # di/dv
+    p = i * v
+    grad_p = v * grad + i  # dp/dv
+    grad2i = -c / nnsvt
+    grad2v = -grad2i * rs
+    grad2p = (
+        grad_v * grad + v * (grad2i/grad_v - grad_i*grad2v/grad_v**2) + grad_i
+    )
+    # generate exact values
+    data = dict(zip((il, io, rs, rsh, nnsvt), ARGS))
+    vdtest = np.linspace(0, pvsystem.estimate_voc(IL, I0, NNSVTH), IVCURVE_NPTS)
+    expected = []
+    for test in vdtest:
+        data[vd] = test
+        test_data = {
+            'i': np.float64(i.evalf(subs=data)),
+            'v': np.float64(v.evalf(subs=data)),
+            'p': np.float64(p.evalf(subs=data)),
+            'grad_i': np.float64(grad_i.evalf(subs=data)),
+            'grad_v': np.float64(grad_v.evalf(subs=data)),
+            'grad': np.float64(grad.evalf(subs=data)),
+            'grad_p': np.float64(grad_p.evalf(subs=data)),
+            'grad2p': np.float64(grad2p.evalf(subs=data))
+        }
+        LOGGER.debug(test_data)
+        expected.append(test_data)
+    return pd.DataFrame(expected, index=vdtest)
+
+
+def test_numerical_precision():
+    """
+    Test that there are no numerical errors due to floating point arithmetic.
+    """
+    expected = pd.read_csv(DATA_PATH)
+    vdtest = np.linspace(0, pvsystem.estimate_voc(IL, I0, NNSVTH), IVCURVE_NPTS)
+    results = pvsystem.bishop88(vdtest, *ARGS, gradients=True)
+    assert np.allclose(expected['i'], results[0])
+    assert np.allclose(expected['v'], results[1])
+    assert np.allclose(expected['p'], results[2])
+    assert np.allclose(expected['grad_i'], results[3])
+    assert np.allclose(expected['grad_v'], results[4])
+    assert np.allclose(expected['grad'], results[5])
+    assert np.allclose(expected['grad_p'], results[6])
+    assert np.allclose(expected['grad2p'], results[7])
+
+
+if __name__ == '__main__':
+    expected = generate_numerical_precision()
+    expected.to_csv(DATA_PATH)
+    test_numerical_precision()
```

Comment:
If the test is not required, I am happy to remove it. I don't really know how it fits in with #426. I apologize for my ignorance.

[I added this "numerical precision" test, very early in this PR](#issuecomment-361502144) in response to [a comment](#issuecomment-360990193) by @adriesse and [another comment](#issuecomment-360995328) by @thunderfish24 regarding accuracy or precision of the explicit calculations.

Perhaps I misunderstood their comments, but all this test does is show that there are no floating point errors in the explicit calculations. [Comments](#issuecomment-361804636) by @wholmgren and others seemed to indicate to me that this test resolved the issue of numerical precision and accuracy.

I have added a module docstring to try to explain that SymPy is a computer algebra system that uses symbolic mathematics like MathCAD, Maple, MuPAD, or Mathematica to represent numbers instead of the standard integer and floating point computer types. Therefore SymPy expressions when evaluated have infinite precision, and are not limited to the machine precision of the computer. The docstring also explains that running the file from the command line genreates a CSV containng high-precision test values (already including in this PR) that the test uses.

The numerical precision test shows that the I-V curve points calculated using  the explicit single diode method `singlediode_methods.bishop88` are within the prescribed tolerance of exactly calculated values with infinite precision (as if you did them by hand, and evaluated each calculation to as many decimal points as you desired).

Reply:
Ok, let's keep this test. I think my confusion came out of some of the discussion in #411 and my mind locked into the wrong idea.